Library/src/Systems/Assets/AssetSystem.cpp

```diff
@@ -849,22 +861,282 @@ CSP_ASYNC_RESULT_WITH_PROGRESS void
 
 void AssetSystem::CreateMaterial(const csp::common::String& Name, const csp::common::String& SpaceId, GLTFMaterialResultCallback Callback)
 {
+	// 1. Create asset collection
+	auto CreateAssetCollectionCB = [this, &Name, Callback](const AssetCollectionResult& CreateAssetCollectionResult)
+	{
+		if (CreateAssetCollectionResult.GetResultCode() != EResultCode::Success)
+		{
+			Callback(GLTFMaterialResult(CreateAssetCollectionResult.GetResultCode(), CreateAssetCollectionResult.GetHttpResultCode()));
+			return;
+		}
+
+		// 2. Create asset
+		const AssetCollection& CreatedAssetCollection = CreateAssetCollectionResult.GetAssetCollection();
+
+		auto CreateAssetCB = [this, Callback, CreatedAssetCollection](const AssetResult& CreateAssetResult)
+		{
+			if (CreateAssetResult.GetResultCode() != EResultCode::Success)
+			{
+				Callback(GLTFMaterialResult(CreateAssetResult.GetResultCode(), CreateAssetResult.GetHttpResultCode()));
+				return;
+			}
+
+			// 3. Upload default material
+			GLTFMaterial NewMaterial;
+			csp::common::String MaterialJson = csp::json::JsonSerializer::Serialize(NewMaterial);
+
+			auto UploadMaterialCallback = [this, Callback, NewMaterial](const UriResult& UploadResult)
+			{
+				if (UploadResult.GetResultCode() != EResultCode::Success)
+				{
+					Callback(GLTFMaterialResult(UploadResult.GetResultCode(), UploadResult.GetHttpResultCode()));
+					return;
+				}
+
+				// 4. Return created material
+				GLTFMaterialResult FinalResult(UploadResult.GetResultCode(), UploadResult.GetHttpResultCode());
+				FinalResult.SetGLTFMaterial(NewMaterial);
+
+				Callback(FinalResult);
+			};
+
+			const Asset& CreatedAsset = CreateAssetResult.GetAsset();
+
+			BufferAssetDataSource AssetData;
+			AssetData.SetMimeType("application/json");
+			AssetData.Buffer	   = const_cast<char*>(MaterialJson.c_str());
```

Comment:
[fix] Hesitant to sign off on const_cast, is this the standard pattern? 

This seems vulnerable to refactoring errors, as this only works coincidentally so long as `MaterialJson` remains alive in the same scope, and code-changes to `BufferAssetDataSource` may corrupt `String` when it comes to destruct.
A `move` is the correct semantic operation here, but given our types arn't there yet, i'd advise doing an explicit copy, so `Buffer` is the sole owning pointer.

Alternately, you could alter `BufferAssetDataSource` to be constructable from a `String&&`, but I suspect that's a big job given that `String` isn't move-aware. :(

Reply:
We do this in other places, unfortunately. I wonder if buffer can be const. I will see what I can do

Reply:
i decided to copy to a string first. Making the variable const was breaking the wrapper gen

Reply:
Thanks for doing that.

[Comment] Just a note though, the comment says that the copy prevents const_cast, which it does, but I'd argue the main virtue of the copy is preventing two objects having owning handles to the same data.